C++/source/dynamic_programming/longest_increasing_subsequence.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+   Longest Increasing Subsequence
+   ------------------------------
+   Given an unsorted array of integers, return the longest increasing subsequence
+
+   Time complexity
+   --------------
+   O(N), where N is size of array
+
+   Space complexity
+   ---------------
+   O(N), where N is size of array
+ */
+
+#include <iostream>
+#include <vector>
+#include <climits>
+#include <algorithm>
+
+using namespace std;
+
+int find_prev_smaller(const vector<int> &a, int x) {
+    for(int i = x-1; i >= 0; i--) {
+        if(a[i] < a[x]) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+int index_of_max(const vector<int> &a) {
+    int max = INT_MIN;
+    int max_i = 0;
+    for(int i = 0; i < (int)a.size(); i++) {
+        if(a[i] > max) {
+            max = a[i];
+            max_i = i;
+        }
+    }
+    return max_i;
+}
+
+vector<int> longest_increasing_subsequence(const vector<int> &a) {
+    vector<int> lis_count(a.size());
+    vector<int> lis_prev(a.size());
+    vector<int> ans;
```

Comment:
Declare variables just before they're used for the first time (which would be line 80 in case of `ans`), and not at the very top of a function.